src/native/cipher.rs

```diff
@@ -261,3 +258,305 @@ pub fn decrypt_in<P: PoolParams>(eta:Num<P::Fr>, memo:&[u8], params:&P)->Vec<Opt
     }
 }
+
+/// get encrypted memo chunks with associated decryption keys (chunk: account or note)
+/// returns vector of tupple (index, chunk, key)
+/// indexes are zero-based and enumerated within current memo
+pub fn symcipher_decryption_keys<P: PoolParams>(eta:Num<P::Fr>, mut memo:&[u8], params:&P) -> Option<Vec<(u64, Vec<u8>, Vec<u8>)>> {
+    let num_size = constants::num_size_bits::<P::Fr>()/8;
+    let account_size = constants::account_size_bits::<P::Fr>()/8;
+    let note_size = constants::note_size_bits::<P::Fr>()/8;
+
+    let nozero_items_num = u32::deserialize(&mut memo).ok()? as usize;
+    if nozero_items_num == 0 {
+        return None;
+    }
+
+    let nozero_notes_num = nozero_items_num - 1;
+    let shared_secret_ciphertext_size = nozero_items_num * constants::U256_SIZE + constants::POLY_1305_TAG_SIZE;
+
+    let account_hash = Num::deserialize(&mut memo).ok()?;
+    let note_hashes = buf_take(&mut memo, nozero_notes_num * num_size)?;
+
+    let shared_secret_text = {
+        let a_p = EdwardsPoint::subgroup_decompress(Num::deserialize(&mut memo).ok()?, params.jubjub())?;
+        let ecdh = a_p.mul(eta.to_other_reduced(), params.jubjub());
+        let key = {
+            let mut x: [u8; 32] = [0; 32];
+            ecdh.x.serialize(&mut &mut x[..]).unwrap();
+            keccak256(&x)
+        };
+        let ciphertext = buf_take(&mut memo, shared_secret_ciphertext_size)?;
+        symcipher_decode::<SHARED_SECRETS_HEAPLESS_SIZE>(&key, ciphertext)
+    };
+
+    if let Some(shared_secret_text) = shared_secret_text {
+        // here is a our transaction, we can restore account and all notes
+        let mut shared_secret_text_ptr = shared_secret_text.as_slice();
+
+        let account_key= <[u8;constants::U256_SIZE]>::deserialize(&mut shared_secret_text_ptr).ok()?;
+        let note_key = (0..nozero_notes_num).map(|_| <[u8;constants::U256_SIZE]>::deserialize(&mut shared_secret_text_ptr)).collect::<Result<Vec<_>,_>>().ok()?;
+
+        let account_ciphertext = buf_take(&mut memo, account_size+constants::POLY_1305_TAG_SIZE)?;
+        let account = decrypt_account(&account_key, account_ciphertext, params)?;
+
+
+        if account.hash(params) == account_hash {
+            let account_tuple = (0 as u64, account_ciphertext.to_vec(), account_key.to_vec());
+            let result = Some(account_tuple)
+                .into_iter()
+                .chain(
+                    (0..nozero_notes_num).filter_map(|i| {
+                    buf_take(&mut memo, num_size)?;
+                    let ciphertext = buf_take(&mut memo, note_size+constants::POLY_1305_TAG_SIZE)?;
+                    let note = decrypt_note(&note_key[i], ciphertext, params)?;
+
+                    let note_hash = {
+                        let note_hash = &mut &note_hashes[i * num_size..(i + 1) * num_size];
+                        Num::deserialize(note_hash).ok()?
+                    };
+
+                    if note.hash(params) != note_hash {
```

Comment:
What do you think about moving this check to decrypt_note? It seems that it doesn't make sense to decrypt notes without this check.

Reply:
Yes, I moved this checking into the decryption routine. The same checking for the account were also moved

https://github.com/zkBob/libzeropool/pull/9/commits/a47890b368d10232e90880fab6f6ef89da53eb9d

Reply:
I forgot that wasm lib use pure decrypt routines to validate compliance report. So I added routines without validation in https://github.com/zkBob/libzeropool/pull/9/commits/80222e57239223a3176ec09f909ebe92f00e99eb Anyway making validation inside a decrypt function simplify program code